### PR TITLE
Add caching to deprecation notice.

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Adds caching to api checks for whether a given file is deprecated

Before
----------------------------------------
Every api call results in multiple file locations (civicrm/api + civicrm/api4 + all extensions being checked for the presences of a file for that entity

After
----------------------------------------
The results are stored in the metadata cache & the filesystem is only interrogated if not set.

The metadata cache uses array storage backed up by Redis or Sql & is fairly long lived - all of which seems appropriate for something slow-changing but frequently accessed

Technical Details
----------------------------------------
We have been seeing a lot of php cycles looking for  & not finding files & so I'm looking at where these might be
addressable.

In this case we can expect that every api call every directory on the path will be checked for the
existing of the relevent file - e.g CustomGroup.php. We really don't need to check for deprecation updates that often.

Comments
----------------------------------------

Our tech ops guy got a list of all the file misses for a short burst while our server was going full throttle - here is the top part - it keeps going on like that through every extension. I'm still looking for other places in the code doing it.

    855 /webroot/civicrm/ext/api4//api/Exception.php
    760 /webroot/civicrm/ext/api4//api/v3/CustomGroup.php
   1140 /webroot/civicrm/ext/api4//api/v3/EntityTag.php
    190 /webroot/civicrm/ext/api4//api/v3/Generic.php
   1045 /webroot/civicrm/ext/api4//api/v3/utils.php
    190 /webroot/civicrm/ext/api4//CRM/Core/BAO/CustomGroup.php
    285 /webroot/civicrm/ext/api4//CRM/Core/BAO/EntityTag.php
    855 /webroot/civicrm/ext/api4//CRM/Core/ClassLoader.php
    855 /webroot/drupal/./api/Exception.php
    760 /webroot/drupal/./api/v3/CustomGroup.php
   1140 /webroot/drupal/./api/v3/EntityTag.php
    190 /webroot/drupal/./api/v3/Generic.php
   1045 /webroot/drupal/./api/v3/utils.php
    190 /webroot/drupal/./CRM/Core/BAO/CustomGroup.php
    285 /webroot/drupal/./CRM/Core/BAO/EntityTag.php
    855 /webroot/drupal/./CRM/Core/ClassLoader.php
    855...extrdir//afform/core//api/Exception.php
    760...extrdir//afform/core//api/v3/CustomGroup.php
   1140...extrdir//afform/core//api/v3/EntityTag.php
    190...extrdir//afform/core//api/v3/Generic.php
   1045...extrdir//afform/core//api/v3/utils.php
    190...extrdir//afform/core//CRM/Core/BAO/CustomGroup.php
    285...extrdir//afform/core//CRM/Core/BAO/EntityTag.php
    855...extrdir//afform/core//CRM/Core/ClassLoader.php
    855...extrdir//afform/gui//api/Exception.php
    760...extrdir//afform/gui//api/v3/CustomGroup.php
   1140...extrdir//afform/gui//api/v3/EntityTag.php
    190...extrdir//afform/gui//api/v3/Generic.php
   1045...extrdir//afform/gui//api/v3/utils.php
    190...extrdir//afform/gui//CRM/Core/BAO/CustomGroup.php
    285...extrdir//afform/gui//CRM/Core/BAO/EntityTag.php
    855...extrdir//afform/gui//CRM/Core/ClassLoader.php
    855...extrdir//afform/html//api/Exception.php
    760...extrdir//afform/html//api/v3/CustomGroup.php
   1140...extrdir//afform/html//api/v3/EntityTag.php
    190...extrdir//afform/html//api/v3/Generic.php
   1045...extrdir//afform/html//api/v3/utils.php
    190...extrdir//afform/html//CRM/Core/BAO/CustomGroup.php
    285...extrdir//afform/html//CRM/Core/BAO/EntityTag.php
    855...extrdir//afform/html//CRM/Core/ClassLoader.php

### CODE REVIEW

**Comment by seamuslee001** [comment]
Looks good to me @eileenmcnaughton

**Comment by eileenmcnaughton** [comment]
@seamuslee001 what about this alternative https://github.com/civicrm/civicrm-core/pull/16012

**Comment by colemanw** [comment]
I'm just wondering why `_civicrm_api3_deprecation_check` does any file scanning at all. If it runs *after* `Entity.get` then the files should already be loaded.

File: api/v3/utils.php [hunk 1]
```diff
@@ -2416,20 +2416,28 @@ function _civicrm_api3_api_resolve_alias($entity, $fieldName, $action = 'create'
...
+ * @return string|array|false
  */
 function _civicrm_api3_deprecation_check($entity, $result = []) {
-  if ($entity) {
+  if (!$entity) {
+    return FALSE;
+  }
+  $cacheKey = 'deprecation_check_entity_' . $entity;
+  if (!Civi::cache('metadata')->has($cacheKey)) {
     $apiFile = "api/v3/$entity.php";
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: @colemanw as in just remove these lines?

Comment (colemanw) [reply api/v3/utils.php]: As in #16460

**Comment by eileenmcnaughton** [comment]
Closing as #16460 is now merged

**Pull request closed without merging**